File: leetcode/363.cc
#include <vector>
#include <cassert>
#include <limits>


using namespace std;


class Solution {
  public:
    int maxSumSubmatrix(vector<vector<int>>& matrix, int k) {
        int r = matrix.size();
        assert(r > 0);
        int c = matrix[0].size();
        assert(c > 0);

        vector<vector<int>> acc(r+1, vector<int>(c+1));
        for (int i = 1; i <= r; ++i) {
            for (int j = 1; j <= c; ++j) {
                acc[i][j] = matrix[i-1][j-1];
                acc[i][j] += acc[i-1][j] + acc[i][j-1] - acc[i-1][j-1];
            }
        }

        int result = numeric_limits<int>::min();
        for (int y1 = 0; y1 <= r; ++y1) {
            for (int y2 = y1+1; y2 <= r; ++y2) {
                for (int x1 = 0; x1 <= c; ++x1) {
                    for (int x2 = x1+1; x2 <= c; ++x2) {
                        int sum = acc[y2][x2] - acc[y1][x2] - acc[y2][x1] + acc[y1][x1];
                        if (sum > k)
                            continue;
                        result = max(result, sum);
                        if (result == k)
                            break;
                    }
                }
            }
        }
        return result;
    }
};

File: leetcode/363.py
from sortedcontainers import SortedList


class Solution:
    def maxSumSubmatrix(self, matrix: List[List[int]], k: int) -> int:
        n = len(matrix)
        assert n > 0
        m = len(matrix[0])
        assert m > 0

        row_accs = []
        for row in matrix:
            cur = [0]
            for c in row:
                cur.append(c + cur[-1])
            row_accs.append(cur)

        result = float('-inf')
        for l in range(m+1):
            for r in range(l+1, m+1):
                row_sums = [row_accs[i][r] - row_accs[i][l] for i in range(n)]

                sl = SortedList([0])
                acc = 0
                for v in row_sums:
                    acc += v
                    t = sl.bisect_left(acc - k)
                    if t < len(sl):
                        result = max(result, acc - sl[t])
                    sl.add(acc)
        return result
